csrc/id_graphs.h

```diff
@@ -0,0 +1,465 @@
+  // sub-regrion of the IdGraph provided. Only that sub-region will be visited.
+  IdGraphVisitor(
+      IdGraph& id_graph,
+      const VectorOfUniqueEntries<IterDomain*> sub_selection = {})
+      : id_graph_(id_graph), sub_selection_(sub_selection) {}
+
+  virtual void handle(IdGroup id_group) = 0;
+  virtual void handle(ExprGroup expr_group) = 0;
+
+  void traverse();
+
+  IdGraph& graph() {
+    return id_graph_;
+  };
+
+  IdGraphVisitor() = delete;
+
+  IdGraphVisitor(const IdGraphVisitor& other) = default;
+  IdGraphVisitor& operator=(const IdGraphVisitor& other) = default;
+
+  IdGraphVisitor(IdGraphVisitor&& other) = default;
+  IdGraphVisitor& operator=(IdGraphVisitor&& other) = default;
+
+  virtual ~IdGraphVisitor() = default;
+
+ private:
+  IdGraph& id_graph_;
+  const VectorOfUniqueEntries<IterDomain*> sub_selection_;
+};
+
+// Statement sorting based on IdGraphVisitor, see warnings to IdGraph Visitor.
+class IdGraphStmtSort : public IdGraphVisitor {
+ public:
+  IdGraphStmtSort(
+      IdGraph& id_graph,
+      const VectorOfUniqueEntries<IterDomain*> sub_selection = {})
+      : IdGraphVisitor(id_graph, sub_selection) {
+    IdGraphVisitor::traverse();
+  }
+
+  ExprGroups exprs() {
+    return sorted_exprs;
+  }
+
+  IdGroups ids() {
+    return sorted_ids;
+  }
+
+  ~IdGraphStmtSort() override = default;
+
+ protected:
+  using IdGraphVisitor::handle;
+  void handle(IdGroup id_group) override {
+    sorted_ids.pushBack(id_group);
+  }
+
+  void handle(ExprGroup expr_group) override {
+    sorted_exprs.pushBack(expr_group);
+  }
+
+  ExprGroups sorted_exprs;
+  IdGroups sorted_ids;
+};
+
+// There's three modes of these iter domain mappings all uniquely important in
+// the lowering process.
+//
+// For EXACT/PERMISSIVE mode consider:
+//
+// consumer[i0, b1] = producer[i0]
+// consumer->merge(0) (consumer will now be [i0 * b1])
+// When producer is replayed as consumer (the direction we use for mapping)
+// with BestEffortReplay forward_bcast_mismatch = True the producer to
+// consumer map will have both a mapping of consumer(i0) to producer(i0) as
+// well as consumer(i0*b1) to producer(i0). This latter mapping is important
+// for loop nest mappings as the consumer will generate a loop based on i0*b1
+// and the producer may be computeAt inside this loop nest. However, for
+// indexing we do not want these two maps as producer may be indexed as i0*i1
+// depending on the loop nest structure and how it was built. Therefore we
+// really need to carry (at least) two sets of maps around for lowering.
+//
+// LOOP mode is important if we have something like:
+// consumer[i0o, threadIdx.x{i0i}] = producer[i0o, threadIdx.y{i0i}](computeAt
+// = 1) which can easily happen when using shared memory. We want to make sure
+// that the iteration domain used for loop construction (concreteId) has the
+// proper parallelization strategy. In parallel mode we do typical iteration
```

Comment:
Do you mean "in loop mode"? I remember we used to have the parallel map.

Reply:
I think I just copied this comment over from compute_at_map.h. So yeah. Though the loop graph is different here, but that's not finalized yet.